experimental/sample_web_static/main.c

```diff
@@ -0,0 +1,187 @@
+void cleanup_sample(iree_sample_state_t* state);
+
+int run_sample(iree_sample_state_t* state, float* image_data);
+
+//===----------------------------------------------------------------------===//
+// Implementation
+//===----------------------------------------------------------------------===//
+
+extern iree_status_t create_device_with_static_loader(
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device);
+
+typedef struct iree_sample_state_t {
+  iree_runtime_instance_t* instance;
+  iree_hal_device_t* device;
+  iree_runtime_session_t* session;
+  iree_vm_module_t* module;
+  iree_runtime_call_t call;
+} iree_sample_state_t;
+
+void iree_sample_state_initialize(iree_sample_state_t* out_state) {
+  out_state->instance = NULL;
+  out_state->device = NULL;
+  out_state->module = NULL;
+  out_state->session = NULL;
+}
+
+iree_status_t create_bytecode_module(iree_vm_module_t** out_module) {
+  const struct iree_file_toc_t* module_file_toc = iree_static_mnist_create();
+  iree_const_byte_span_t module_data =
+      iree_make_const_byte_span(module_file_toc->data, module_file_toc->size);
+  return iree_vm_bytecode_module_create(module_data, iree_allocator_null(),
+                                        iree_allocator_system(), out_module);
+}
+
+iree_sample_state_t* setup_sample() {
+  iree_sample_state_t* state;
+  state = malloc(sizeof(iree_sample_state_t));
+  iree_sample_state_initialize(state);
+
+  iree_runtime_instance_options_t instance_options;
+  iree_runtime_instance_options_initialize(IREE_API_VERSION_LATEST,
+                                           &instance_options);
+  // Note: no call to iree_runtime_instance_options_use_all_available_drivers().
+
+  iree_status_t status = iree_runtime_instance_create(
+      &instance_options, iree_allocator_system(), &state->instance);
+
+  if (iree_status_is_ok(status)) {
+    status = create_device_with_static_loader(iree_allocator_system(),
+                                              &state->device);
+  }
+
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_session_create_with_device(
+        state->instance, &session_options, state->device,
+        iree_runtime_instance_host_allocator(state->instance), &state->session);
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = create_bytecode_module(&state->module);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_session_append_module(state->session, state->module);
+  }
+
+  const char kMainFunctionName[] = "module.predict";
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_call_initialize_by_name(
+        state->session, iree_make_cstring_view(kMainFunctionName),
+        &state->call);
+  }
+
+  if (!iree_status_is_ok(status)) {
+    iree_status_fprint(stderr, status);
+    iree_status_free(status);
+    free(state);
```

Comment:
you should be calling cleanup_sample here (or really, `iree_sample_state_deinitialize` to match the `iree_sample_state_initialize`) - right now this is leaking the session and instance (at least)

it's always best to zero everything in the initializer and then have the cleanup code be tolerant to any individual thing being zero (release/etc are) - that way you can reuse the deinit as cleanup for failed inits.

Reply:
Ah, good catch. The failure path should indeed clean things up back to a consistent state.

On a related note, I was thinking about calling the exposed `cleanup_sample()` function from JavaScript on page suspend (or after a period of inactivity), rather than between invocations. It isn't call at all right now, which "leaks" as well. I wasn't quite sure what would be best when bridging between a C API with explicit memory management (and potentially costly startup/initialization methods) and a web app.